plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/CaseReduce.hs

```diff
@@ -13,16 +15,19 @@ import Control.Lens (transformOf)
 import Data.Vector qualified as V
 
 caseReduce
-    :: Monad m
+    :: (Monad m, CaseBuiltin (Term name uni fun a) uni)
     => Term name uni fun a
     -> SimplifierT name uni fun a m (Term name uni fun a)
 caseReduce term = do
     let result = transformOf termSubterms processTerm term
     recordSimplification term CaseReduce result
     return result
 
-processTerm :: Term name uni fun a -> Term name uni fun a
+processTerm :: CaseBuiltin (Term name uni fun a) uni => Term name uni fun a -> Term name uni fun a
 processTerm = \case
     Case ann (Constr _ i args) cs | Just c <- (V.!?) cs (fromIntegral i) ->
                                     mkIterApp c ((ann,) <$> args)
-    t                                                     -> t
+    Case ann (Constant _ con) cs -> case caseBuiltin con cs of
+        Left _    -> Error ann
```

Comment:
Worth a comment: why we rewrite it to `Error` here, but not in the `Constr` case.

Reply:
I... don't know. I just thought that if it was an error, I can make it `Error` to save some size.

I can make the `Constr` case rewrite it to `Error` as well if you're fine with that.

Reply:
Hm, I'm not sure how much it matters, but I don't mind punishing someone who created an invalid script by not optimizing it fully. OK, I'll remove the `Error` part. Thanks!